[7.x] require master node for watcher template (#69147)

jakelandis opened the pull request:
Backports the following commits to 7.x:
 - require master node for watcher template (#69147)

jakelandis committed:
require master node for watcher template (#69147)

This commit requires Watcher to install composable index templates
only from the master node since during a rolling upgrade it could
try to talk to a master node that predates composable index templates.

Some tests have also been adjusted to that allows for this change.
Those test changes have been discussed offline and can generally be
summarized as the things they are testing is an edge case where if that
edge case (breaking changes to the mappings of watcher history index) between
versions then there would be other concerns beyond the need for which
node installs the watcher templates.

closes #66837

Co-authored-by: Martijn van Groningen <martijn.v.groningen@gmail.com>
# Conflicts:
#	x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/WatcherIndexTemplateRegistry.java
#	x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/support/WatcherIndexTemplateRegistryTests.java

jakelandis committed:
checkstyle

jakelandis commented:
@martijnvg @probakowski - could you help here with the backport ?  Due to a merge conflict in the `validate` method, I did not include the change from master here on the backport. I wasn't sure if the [change](https://github.com/elastic/elasticsearch/pull/69147/commits/73a422e62db0ea68d747fa89c997d0580c6b0d6f) @martijnvg applied to master is relevant here.

martijnvg commented:
I think we need to have a similar change to the `validate` method in this pr too, but then with different versions. 
I will look into this.

martijnvg commented:
It looks like the last version used in 6.x line is version 9. I don't see any breaking changes to the watcher history template after that. So I think that in the `WatcherIndexTemplateRegistry#validate(...)` method we should check for whether a watcher history template exists and that the version should be at least 6.

martijnvg committed:
having a history index template from version 6 is okay

martijnvg committed:
fixed checkstyle violation

jakelandis commented:
thanks @martijnvg ! 

@probakowski - would mind reviewing ?

probakowski commented on x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/WatcherIndexTemplateRegistry.java:
```diff
@@ -108,18 +110,23 @@ protected String getOrigin() {
         }
+        return watcherHistoryTemplateIds.filter(s -> s.startsWith(".watch-history-"))
+            .map(s -> Integer.valueOf(s.substring(s.lastIndexOf('-') + 1)))
+            .anyMatch(version -> version >= 6);
```
Should this be `9` and not `6`? Last `6.x` uses template version `9`, right?

martijnvg replied:
Oops, yes version 9 is the version used in 6.8.latest, I misread the `WatcherIndexTemplateRegistryField` in the 6.8 branch. Although looking at the history I don't think even breaking changes were introduced from template version 6.
In anyway I will update this version to 9, since we only support rolling upgrades from 6.8.x to 7.x.

probakowski reviewed (commented):
Overall looks good, just one question around version used

martijnvg committed:
update version check,

ensure that at least watcher history template version 9 is installed before starting watcher.

probakowski reviewed (approved):
LGTM 👍

martijnvg commented:
I'm also backporting this change to the 7.12 branch: #69855

jakelandis commented:
Thanks @martijnvg and @probakowski !
